File: ALGO/python/GRAPHS/KOSARAJUS/code.py
from collections import defaultdict


class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = defaultdict(list)

    def addEdge(self, u, v):
        self.graph[u].append(v)

    def DFS(self, v, visited):
        visited[v] = True
        print(v,end=" ")
        for x in self.graph[v]:
            if(visited[x] == False):
                self.DFS(x, visited)

    def FillOrder(self, v, visited, stack):
        visited[v] = True
        for x in self.graph[v]:
            if(visited[x] == False):
                self.FillOrder(x,visited,stack)
        stack = stack.append(v)

    def getTranspose(self):
        g = Graph(self.V)
        for x in self.graph:
            for y in self.graph[x]:
                g.addEdge(y, x)
        return g

    def printStronglyConnected(self):
        stack = []
        visited = [False]*(self.V)
        for i in range(self.V):
            if(visited[i] == False):
                self.FillOrder(i, visited, stack)
        gr = self.getTranspose()
        visited = [False]*(self.V)
        while(stack):
            i = stack.pop()
            if(visited[i] == False):
                gr.DFS(i, visited)
            print()


g = Graph(5)

g.addEdge(1, 0)

g.addEdge(0, 2)

g.addEdge(2, 1)

g.addEdge(0, 3)

g.addEdge(3, 4)

print ("Following are strongly connected components " "in given graph")

g.printStronglyConnected()

File: 2.DataStructure/C++/GRAPHS/KOSARAJU/code.cpp
// # KOSARAJUs ALGORITHM FOR STRONGLY CONNECTED GRAPHS

#include <bits/stdc++.h>
using namespace std;

class Graph
{
    int V;
    list<int> *adj;

public:
    Graph(int V_)
    {
        this->V = V_;
        adj = new list<int>[V];
    }
    void addEdge(int src, int dest)
    {
        adj[src].push_back(dest);
    }
    void FillStackOrder(int v, bool *visited, stack<int> &Stack)
    {
        visited[v] = true;
        for (auto it : adj[v])
        {
            if (!visited[it])
            {
                FillStackOrder(it,visited, Stack);
            }
        }
        Stack.push(v);
    }
    Graph getTranspose()
    {
        Graph g(V);
        for (int i = 0; i < V; i++)
        {
            for (auto it : adj[i])
            {
                g.adj[it].push_back(i);
            }
        }
        return g;
    }
    void DFS(int v, bool *visited)
    {
        visited[v] = true;
        cout << v << " ";
        for (auto it : adj[v])
        {
            if (!visited[it])
            {
                DFS(it, visited);
            }
        }
    }
    void printStronglyConnected()
    {
        stack<int> Stack;
        bool *visited = new bool[V];
        memset(visited, 0, sizeof visited);
        for (int i = 0; i < V; i++)
        {
            if (!visited[i])
            {
                FillStackOrder(i,visited,Stack);
            }
        }
        Graph gr = getTranspose();
        memset(visited, 0, sizeof visited);
        while (Stack.empty() == false)
        {
            int v = Stack.top();
            Stack.pop();
            if (!visited[v])
            {
                gr.DFS(v, visited);
                cout << endl;
            }
        }
    }
};

int main()
{
    Graph g(5);

    g.addEdge(1, 0);

    g.addEdge(0, 2);

    g.addEdge(2, 1);

    g.addEdge(0, 3);

    g.addEdge(3, 4);

    cout << "Following are strongly connected components in "

            "given graph \n"; 

    g.printStronglyConnected();
    return 0;
}